qiskit/transpiler/passes/optimization/template_matching/template_substitution.py

```diff
@@ -0,0 +1,346 @@
+    """
+    Class to store the configuration of a given match substitution, which circuit
+    gates, template gates, qubits, and clbits and predecessors of the match
+    in the circuit.
+    """
+
+    def __init__(self, circuit_config, template_config, pred_block,
+                 qubit_config, clbit_config=None):
+        self.circuit_config = circuit_config
+        self.template_config = template_config
+        self.qubit_config = qubit_config
+        self.clbit_config = clbit_config if clbit_config is not None else []
+        self.pred_block = pred_block
+
+
+class TemplateSubstitution:
+    """
+    Class to run the subsitution algorithm from the list of maximal matches.
+    """
+
+    def __init__(self, max_matches, circuit_dag_dep, template_dag_dep):
+        """
+        Initialize TemplateSubstitution with necessary arguments.
+        Args:
+            max_matches (list): list of maximal matches obtained from the running
+             the template matching algorithm.
+            circuit_dag_dep (DAGDependency): circuit in the dag dependency form.
+            template_dag_dep (DAGDependency): template in the dag dependency form.
+        """
+
+        self.match_stack = max_matches
+        self.circuit_dag_dep = circuit_dag_dep
+        self.template_dag_dep = template_dag_dep
+
+        self.substitution_list = []
+        self.unmatched_list = []
+        self.dag_dep_optimized = DAGDependency()
+        self.dag_optimized = DAGCircuit()
+
+    def _pred_block(self, circuit_sublist, index):
+        """
+        It returns the predecessors of a given part of the circuit.
+        Args:
+            circuit_sublist (list): list of the gates matched in the circuit.
+            index (int): Index of the group of matches.
+        Returns:
+            list: List of predecessors of the current match circuit configuration.
+        """
+        predecessors = set()
+        for node_id in circuit_sublist:
+            predecessors = predecessors | set(self.circuit_dag_dep.get_node(node_id).predecessors)
+
+        exclude = set()
+        for elem in self.substitution_list[:index]:
+            exclude = exclude | set(elem.circuit_config) | set(elem.pred_block)
+
+        pred = list(predecessors - set(circuit_sublist) - exclude)
+        pred.sort()
+
+        return pred
+
+    def _rules(self, circuit_sublist):
+        """
+        Set of rules to decide whether the match is to be substitute or not.
+        Args:
+            circuit_sublist (list): list of the gates matched in the circuit.
+        Returns:
+            bool: True if the match respects the given rule for replacement, False otherwise.
+        """
+        if len(circuit_sublist) > (self.template_dag_dep.size() / 2):
+            for elem in circuit_sublist:
+                for config in self.substitution_list:
+                    if any(elem == x for x in config.circuit_config):
+                        return False
+            return True
```

Comment:
this is the part where you decide to substitute or not, right? is it purely based on whether the match is more than half the template or not? what if this half is the cheap half (like it is all CNOTs and the other half is all Toffolis)?

Reply:
It was the first version. I added comparison of quantum costs afterwards. It will be in the update.